Adds elementary gate decomposition for single-qubit `QubitUnitary`

glassnotes opened the pull request:
**Context:** Currently, there is no decomposition/synthesis for the arbitrary `QubitUnitary` operation.

**Description of the Change:** Adds the math needed to express arbitrary single-qubit matrices as a `Rot` gate (or an `RZ` gate, if they are diagonal).

**Benefits:** This should allow `QubitUnitary` to run on more devices. Furthermore, I think that by using the decomposition we can make `QubitUnitary` fully differentiable, at least for the single-qubit case (see example below).

**Possible Drawbacks:** None

**Related GitHub Issues:** None

**Example:**

```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device('default.qubit', wires=1)

def qfunc(x):
    qml.Hadamard(wires=0)

    # Create a unitary matrix based on an input parameter
    Ux = np.array([[1, 0], [0, np.exp(1j * x)]])
    ops = qml.QubitUnitary.decomposition(Ux, wires=0)
    for op in ops:
        op.queue()

    return qml.expval(qml.PauliX(0))
```
```pycon
>>> original_qnode = qml.QNode(qfunc, dev, diff_method="adjoint")
>>> print(qml.grad(original_qnode)(0.2))
-0.19866933079506127
```

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -2157,6 +2157,43 @@ def _matrix(cls, *params):
+                    qml.math.real(phi), qml.math.real(theta), qml.math.real(omega), wires=wires[0]
+                )
+            ]
+        else:
+            return NotImplementedError
```
```suggestion
        return NotImplementedError("Decompositions only supported for single-qubit unitaries")
```

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -2157,6 +2157,43 @@ def _matrix(cls, *params):
+    @staticmethod
+    def decomposition(U, wires):
+        # Decompose arbitrary single-qubit unitaries as the form RZ RY RZ
+        if U.shape[0] == 2:
```
```suggestion
        if qml.math.shape(U)[0] == 2:
```

Since not all frameworks *cough* tensorflow *cough* return integers from the `shape` property

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -2157,6 +2157,43 @@ def _matrix(cls, *params):
+            # if desired. If the top left element is 0, can only use the off-diagonal elements
+            if qml.math.isclose(U[0, 0], 0):
+                phi = 1j * qml.math.log(U[0, 1] / U[1, 0])
+                omega = -phi - 2 * qml.math.angle(U[1, 0])
```
You may need to be careful here:

- `phi` is complex
- `2 * angle(...)` will be real (I think `float64`)

and TensorFlow (maybe also PyTorch?) will not automatically cast the latter to a complex number before doing the subtraction.

I think the following should work:

```suggestion
                omega = -phi - qml.math.cast_like(2 * qml.math.angle(U[1, 0]), phi)
```

josh146 replied:
(Note: `qml.math.cast(..., "complex128")` would also work above, but `cast_like` is safer, in that it will support both `complex64` and `complex128` depending on whatever `omega` is)

glassnotes replied:
Oh thanks, yes this looks safer! I originally had the conversion to `real` in the line of `phi` rather than in the return value, do you think it'd be better to convert `phi` first so that the subtract ends up real? Or is it better to just do the casting anyways?

josh146 replied:
I think you'd need the casting regardless, potentially you could have float64 subtract float32, which would also cause an error

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -2157,6 +2157,43 @@ def _matrix(cls, *params):
+                omega = -phi - 2 * qml.math.angle(U[1, 0])
+            else:
+                omega = 1j * qml.math.log(qml.math.tan(theta / 2) * U[0, 0] / U[1, 0])
+                phi = -omega - 2 * qml.math.angle(U[0, 0])
```
```suggestion
                phi = -omega - qml.math.cast_like(2 * qml.math.angle(U[0, 0]), omega)
```

josh146 commented on tests/ops/test_qubit_ops.py:
```diff
@@ -1445,6 +1445,29 @@ def test_qubit_unitary_not_matrix_exception(self, U):
+            (np.exp(1j * 0.02) * qml.Rot(-1, 2, -3, wires=0).matrix, qml.Rot, [-1, 2, -3]),
+        ],
+    )
+    def test_qubit_unitary_decomposition(self, U, expected_gate, expected_params):
+        decomp = qml.QubitUnitary.decomposition(U, wires=0)
+
+        assert len(decomp) == 1
+        assert isinstance(decomp[0], expected_gate)
+        assert np.allclose(decomp[0].parameters, expected_params)
```
Nice! Although, I think important to also add:

- Tests where U is a TF/Torch/JAX tensor
- Differentiation tests for TF/Torch/JAX/autograd

josh146 reviewed (changes requested):
@glassnotes this is awesome! A couple of things that crossed my mind:

- This means that we can now support `qml.Hermitian` (arbitrary Hermitian matrix observables) directly on hardware, since it currently diagonalizes down to a `QubitUnitary` rotation 🙌 

- I love that it results in differentiability 🤯 

- In your code block in the PR comment, I don't think you need the `op.queue()`? Calling `.decomposition` should automatically queue the ops, so they might be being queued twice.

- Part of me wonder if this feels like a compilation transform? Becuase there will be use cases where users _want to force_ a unitary decomposition on a device that natively supports it. But I don't know? it would add overhead compared to simply calling `QubitUnitary.decomposition()` manually.

- Don't forget to add interface tests! To ensure it works with different tensor input

- Along the same lines, gradient tests should also be added 🙂

glassnotes commented:
> Part of me wonder if this feels like a compilation transform? Becuase there will be use cases where users want to force a unitary decomposition on a device that natively supports it. But I don't know? it would add overhead compared to simply calling QubitUnitary.decomposition() manually.

I am wondering this too; so, for additional context, I'd like to do this for the 2-qubit case as well. That's going to make the `decomposition` method extremely hairy. If we use transforms, we can make things more flexible (e.g., we can have different transforms for `ZYZ` versus `ZXZ` decompositions), and better separate the 1- and 2-qubit cases from each other. Then we can have something that passes through the circuit, finds `QubitUnitary`s, and depending on the number of qubits calls the appropriate transform.

josh146 commented:
> I am wondering this too; so, for additional context, I'd like to do this for the 2-qubit case as well. 

Oh! Maybe the answer is... both?

- We have transforms to make it flexible and keep the code organized.
- The `QubitUnitary.decomposition()` method calls the transforms as needed, in order to support devices where `QubitUnitary` is not defined.

glassnotes commented:
>     * We have transforms to make it flexible and keep the code organized.
> 
>     * The `QubitUnitary.decomposition()` method calls the transforms as needed, in order to support devices where `QubitUnitary` is not defined.

I like it! I'll give it a go and we can see what it looks like. After deciding on that I'll make the rest of the tests.

josh146 commented on tests/transforms/test_decompositions.py:
```diff
@@ -0,0 +1,380 @@
+        original_grad = qml.grad(original_qnode)(input)
+        transformed_grad = qml.grad(transformed_qnode)(input)
+
+        assert qml.math.allclose(original_grad, transformed_grad)
```
💪

josh146 commented on tests/transforms/test_decompositions.py:
```diff
@@ -0,0 +1,380 @@
+
+        original_grad = jax.grad(original_qnode)(input)
+        transformed_grad = jax.grad(transformed_qnode)(input)
+        assert qml.math.allclose(original_grad, transformed_grad, atol=1e-7)
```
amazing testing 💯

josh146 reviewed (changes requested):
I really like the new organization! Especially the fact that the decompositions can be done three ways:

- Automatically if the device does not support it
- Manually by calling `qml.QubitUnitary.decomposition(U)`
- Manually on an existing qfunc via a transform

glassnotes commented on pennylane/ops/qubit.py:
```diff
@@ -2157,6 +2157,16 @@ def _matrix(cls, *params):
 
         return U
 
+    @staticmethod
+    def decomposition(U, wires):
+        # Decompose arbitrary single-qubit unitaries as the form RZ RY RZ
+        if qml.math.shape(U)[0] == 2:
+            wire = Wires(wires)[0]
+            decomp_ops = qml.transforms.decompositions._zyz_decomposition(U, wire)
+            return decomp_ops
+
+        return NotImplementedError("Decompositions only supported for single-qubit unitaries")
```
Yes, thanks for the suggestion to separate it out! :grin: It will also make it much cleaner to add different decompositions, one for 2-qubit unitaries, etc. while keeping the qubit ops file tidy.

glassnotes commented on pennylane/transforms/decompositions.py:
```diff
@@ -0,0 +1,114 @@
+    return U
+
+
+def _zyz_decomposition(U, wire):
```
Oh good point, let's make it non-private since it's used outside the file. :+1:

glassnotes commented on pennylane/transforms/decompositions.py:
```diff
@@ -0,0 +1,114 @@
+    """
+    for op in tape.operations + tape.measurements:
+        if isinstance(op, qml.QubitUnitary):
+            dim_U = qml.math.shape(op.parameters[0])[0]
+
+            if dim_U != 2:
+                continue
```
Actually I think the checks here are the ones to disable since this checked in `_zyz_decomposition` via `_convert_to_su2`. I think the check needs to stay in there because that's what gets called by the `QubitUnitary.decomposition` method (which itself doesn't do a full check for unitarity).

glassnotes commented on pennylane/transforms/__init__.py:
```diff
@@ -45,6 +45,7 @@
 
     ~adjoint
     ~ctrl
+    ~transforms.decompose_single_qubit_unitaries
```
Should we include `zyz_decomposition` in the docs here?

josh146 replied:
Sure!

josh146 replied:
We are getting to the size where maybe we should just use `automodapi` rather than hand listing everything?

josh146 replied:
Oh wait, there is an issue, in that some of the functions listed here are top level, and some aren't 😢

glassnotes replied:
On second thought, it doesn't really fit in this category since it is itself not a transform - maybe we leave it out for now, but then include it when the compilation transforms get added (it would fit better with those).

glassnotes commented:
@josh146 Thanks for the review! Have addressed all your comments, and split the decomposition methods into their own submodule, separate from the transform.

josh146 commented on pennylane/transforms/decompose_single_qubit_unitaries.py:
```diff
@@ -0,0 +1,86 @@
+
+        def qfunc():
+            qml.QubitUnitary(U, wires=0)
+            return qml.expval(qml.PauliZ(0)
```
```suggestion
            return qml.expval(qml.PauliZ(0))
```

josh146 commented on pennylane/transforms/decompositions/__init__.py:
```diff
@@ -0,0 +1,18 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""This module contains decompositions for (numerically-specified) arbitrary
+unitary operations into sequences of elementary operations.
```
I can imagine this being really useful going forward! Both for devs, and external contributors that want to contribute new decompositions to PL without the overhead of having to learn the `Operations` class.

josh146 commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -0,0 +1,105 @@
+    r"""Recover the decomposition of a single-qubit matrix :math:`U` in terms of
+    elementary operations.
+
+    Diagonal operations will be converted to a single ``RZ`` gate, while non-diagonal
+    operations will be converted to a ``Rot`` gate that implements the original operation
```
```suggestion
    Diagonal operations will be converted to a single :class:`.RZ` gate, while non-diagonal
    operations will be converted to a :class:`.Rot` gate that implements the original operation
```

josh146 reviewed (changes requested):
All my comments are minor, as soon as the failing tests are fixed, I'm happy to approve!

glassnotes commented on pennylane/transforms/decompose_single_qubit_unitaries.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@qfunc_transform
+def decompose_single_qubit_unitaries(tape):
```
> decompose_single_qubit_unitaries(tape, decomp=zyz_decomposition)

Oh I like this idea!

> A separate thought: I don't really like the name, but am going to be really annoying when I say I don't have any better suggestions 
> 
> I find it a bit too long
> I keep remembering that the compile PR uses 'expand' instead of 'decompose'

Yes I feel the same (about this and some of the other compilation transforms). Some other ideas:
 - `decompose_qubit_unitaries`
 - `unitaries_to_rot`  / `unitary_to_rot` 
 - `qubit_unitary_to_rot`

josh146 replied:
Oh I like `unitary_to_rot`

glassnotes replied:
Me too :+1: It's the shortest and also captures that it's for single-qubit operations :tada:

glassnotes replied:
> decompose_single_qubit_unitaries(tape, decomp=zyz_decomposition)

Actually, if we are going to have the transform convert things to `Rot`, it makes sense to keep as-is rather than adding the option, since `Rot` is intrinsically `zyz`. But later if we add more decompositions and start returning them as `RZ, RY, RZ`, etc. instead of `Rot`, this would make things more flexible.

glassnotes commented on pennylane/transforms/decompose_single_qubit_unitaries.py:
```diff
@@ -0,0 +1,86 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```
I'm inclined to leave it where it is to keep the actual transform (which could be used in a compilation pipeline) separate from the different decompositions, which are not transforms and have multiple use cases.

glassnotes commented on tests/circuit_graph/test_qasm.py:
```diff
@@ -227,12 +227,12 @@ def test_basis_state_initialization_decomposition(self):
-
         with qml.tape.QuantumTape() as circuit:
-            qml.S(wires=0), qml.QubitUnitary(U, wires=[0, 1])
+            qml.S(wires=0), qml.DoubleExcitationPlus(0.5, wires=[0, 1, 2, 3])
```
It was actually hard to find an operation that didn't have a decomposition into QASM gates that I could use here... but this is a good thing :grin:

josh146 commented on tests/transforms/test_unitary_to_rot.py:
```diff
@@ -0,0 +1,318 @@
+
+        original_grad = jax.grad(original_qnode)(input)
+        transformed_grad = jax.grad(transformed_qnode)(input)
+        assert qml.math.allclose(original_grad, transformed_grad, atol=1e-7)
```
Amazing tests!!

josh146 reviewed (approved):
This functionality looks great @glassnotes, very happy to approve! In particular, I like how you separated the tests into tests that _only_ check interface decomposition, and tests that only check interface gradients. It is very tempting when writing tests to combine them into a single test (which I have done a lot in the codebase).

Just two minor comments:

- `zyz_decomposition` doesn't currently appear in the docs
- I'm not sure how important this is, but it could be nice to parametrize the gradient tests with `diff_method = ["parameter-shift", "backprop"]`, _just_ to make sure everything works correctly with the non-default `parameter-shift` logic.

glassnotes commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -0,0 +1,105 @@
+        phi = 1j * math.log(U[0, 1] / U[1, 0])
+        omega = -phi - math.cast_like(2 * math.angle(U[1, 0]), phi)
+    else:
+        el_division = U[0, 0] / U[1, 0]
```
:rofl: :rofl:

glassnotes commented on pennylane/ops/qubit.py:
```diff
@@ -2159,6 +2159,16 @@ def _matrix(cls, *params):
+            decomp_ops = qml.transforms.decompositions.zyz_decomposition(U, wire)
+            return decomp_ops
+
+        raise NotImplementedError("Decompositions only supported for single-qubit unitaries")
```
Not sure why this isn't getting caught; the corresponding test is on line 1593 of the `test_qubit_ops` file.

josh146 replied:
Ah, I think this is a well-known coverage bug. Fine to ignore.

glassnotes commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -0,0 +1,105 @@
+        ])
+
+    For PennyLane devices that cannot natively implement ``QubitUnitary``, we
+    can instead recover a ``Rot`` gate that implements the same operation, up
+    to a global phase:
```
Nope, it only affect the explicit matrix representation and not the outcomes :+1:

glassnotes commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -0,0 +1,105 @@
+    return U
+
+
+def zyz_decomposition(U, wire):
```
I'm actually not sure about this one, because it's not a transform. If we know that this is where a larger set of decompositions is going to live, then it makes sense to have a separate section (or even page) with all that info. But since I went and made it a public method, it should probably live somewhere, so I just added a small subsection to the quantum functions part of the docs.

glassnotes replied:
(BTW using `..automodapi` threw an error; which is why I added it the way I did with `autosummary`.)

josh146 replied:
👍
